[FLINK-10074] Allowable number of checkpoint failures

yanghua opened the pull request:
## What is the purpose of the change

*This pull request allows number of checkpoint failures*

## Brief change log

 - *Add a new API `CheckpointConfig#setTolerableCheckpointFailuresNumber`*
 - *Add a new API `CheckpointConfig#getTolerableCheckpointFailuresNumber`*
- *Changed `FailingCheckpointExceptionHandler#tryHandleCheckpointException`*

## Verifying this change

This change added tests and can be verified as follows:

  - *CheckpointExceptionHandlerTest#testRethrowingHandlerWithTolerableNumberTriggered*
  - *CheckpointExceptionHandlerTest#testRethrowingHandlerWithTolerableNumberNotTriggered*
 
## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

yanghua commented:
cc @tillrohrmann

tillrohrmann commented:
Thanks for opening this PR @yanghua. I'll try to give it a review in the next days.

yanghua commented:
@tillrohrmann When you have free time, please review this PR. thanks.

tweise commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/CheckpointExceptionHandlerTest.java:
```diff
@@ -49,6 +49,59 @@ public void testRethrowingHandler() {
+			failedCheckpointMetaData = new CheckpointMetaData(44L, 4711L);
+			exceptionHandler.tryHandleCheckpointException(failedCheckpointMetaData, testException);
+			failedCheckpointMetaData = new CheckpointMetaData(45L, 4711L);
+			exceptionHandler.tryHandleCheckpointException(failedCheckpointMetaData, testException);
```
Shouldn't it verify that this attempt generated the exception and not previous?

yanghua replied:
here, This line will throw an exception, then it will enter the catch block and verify that the exception is what we expected. If no exception is thrown to this line, then Assert.fail() will be fired, indicating that the actual exception is not thrown and the test will fail. Of course, as you said, this verification does not prove that the exception we need is thrown in this line. Maybe I can try and check each checkpoint independently.

yanghua commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/CheckpointExceptionHandlerTest.java:
```diff
@@ -49,6 +49,59 @@ public void testRethrowingHandler() {
+			failedCheckpointMetaData = new CheckpointMetaData(46L, 4711L);
+			exceptionHandler.tryHandleCheckpointException(failedCheckpointMetaData, testException);
+		} catch (Exception e) {
+			Assert.assertNotEquals(testException, e);
```
Yes, we can just throw it.

azagrebin commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/CheckpointExceptionHandlerConfigurationTest.java:
```diff
@@ -75,6 +79,19 @@ public CheckpointExceptionHandler createCheckpointExceptionHandler(
+			CheckpointExceptionHandlerFactory.FailingCheckpointExceptionHandler actuallyHandler =
+				(CheckpointExceptionHandlerFactory.FailingCheckpointExceptionHandler) exceptionHandler;
+
+			Assert.assertEquals(3, actuallyHandler.tolerableNumber);
```
I would rather make fields of `FailingCheckpointExceptionHandler` private and check `tolerableNumber` the same way as `failTaskOnCheckpointException` in previous declaration of `CheckpointExceptionHandlerFactory`

azagrebin commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/CheckpointExceptionHandlerConfigurationTest.java:
```diff
@@ -63,6 +63,10 @@ private void testConfigForwarding(boolean failOnException) throws Exception {
 		environment.getExecutionConfig().setFailTaskOnCheckpointError(expectedHandlerFlag);
 
+		if (failOnException) {
+			environment.getExecutionConfig().setTaskTolerableCheckpointFailuresNumber(3);
```
this can be always set, as it should be just ignored if `failOnException` is `false`

azagrebin commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/CheckpointExceptionHandlerFactory.java:
```diff
@@ -48,12 +48,41 @@ public CheckpointExceptionHandler createCheckpointExceptionHandler(
+				return true;
+			}
+
+			if (checkpointMetaData.getCheckpointId() - latestFailedCheckpointID == 1) {
```
I think rather than relying on sequential numbering of checkpoints, 
it is better we add one more signal: `CheckpointExceptionHandler.checkpointSucceeded()` where the counter is reset. 

This method can be called in `AsyncCheckpointRunnable.run()`, e.g. after `reportCompletedSnapshotStates` is done:
```
owner.asynchronousCheckpointExceptionHandler.checkpointSucceeded(); // forward it to synchronousCheckpointExceptionHandler inside
```

The checkpoints finish concurrently, so I think we have to use an `AtomicInteger` for the `cpFailureCounter` and `cpFailureCounter.incrementAndGet()`.

yanghua commented:
@azagrebin thanks for your suggestion, I have refactored this PR, please review again~

yanghua commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -875,6 +876,9 @@ public void run() {
 						localTaskOperatorSubtaskStates,
 						asyncDurationMillis);
 
+					if (owner.asynchronousCheckpointExceptionHandler != null) {
```
@azagrebin In the normal runtime it will not be null, plus this judgment is mainly for the checkpoint test code (it does not perform the initialization logic for the asynchronousCheckpointExceptionHandler), if not added then the test code will report to the NPE, so in the second commit, I added this judgment.

azagrebin replied:
I see, I would rather initialise it in `NoOpStreamTask` to avoid `null` but it would need then more refactoring.

yanghua replied:
I looked at the StreamTask, the initialization logic of the asynchronousCheckpointExceptionHandler is in the invoke method and before the call to the concrete init method. The invoke method has been marked as final. So, I personally feel that we don't have to make this thing too complicated, just to reduce a null judgment. What do you think?

azagrebin replied:
ok, I agree that it would need more refactoring to remove this null check.

yanghua commented:
@azagrebin thanks for your suggestion, refactored this PR.

yanghua commented:
Hi @tillrohrmann , Currently it is applied to the sub task instance (TM) instead of the checkpoint coordinator (JM) as a complement to the setFailOnCheckpointingErrors method, since setFailOnCheckpointingErrors only provides the user with yes or no choices.

But it sounds reasonable to apply the count on the Job globally.

yanghua commented:
@tweise Any comments or opinions? The current count implementation is for a single sub task instance, not the job level. Maybe @tillrohrmann  's idea is more reasonable?

klion26 commented:
Thanks for opening this PR @yanghua , as a user, I like the idea fail in JM other than TM

tillrohrmann commented:
I understand why you implemented it the way you did. I think the `setFailOnCheckpointingErrors` should actually also be handled by the `CheckpointCoordinator`/`JM` and not on the Task level (but this is a different story). This looks a little bit like a shortcut we made back in the days.

yanghua commented:
@tillrohrmann I originally wanted to do it in the checkpoint coordinator when I first implemented it, but when I touched setFailOnCheckpointingErrors, I was guided and turned to the TM. I have also thought about this issue recently and it seems that it is more reasonable to implement on the checkpoint coordinator. Well, I will implement it again based on the checkpoint coordinator.

To @klion26 thanks for your opinion, will change the implementation.

tweise commented:
@yanghua @tillrohrmann I think a user would normally expect this count to apply globally, but please also consider the case of an intermittent failure (like S3 rate limit or storage backend unavailable for other reason). In a large job that would cause potentially many subtasks to fail in parallel. While this could be addressed by setting a corresponding very high threshold, it would in turn mean a problem that is isolated to a single task would not hit the threshold until much much later, leaving the job in flipflop status instead of failing.

tillrohrmann commented:
A failure of a global checkpoint should only increment the failure count by one independent of the number of failed subtasks. Thus, I would hope that one does not need to set a different threshold for the two different cases you described @tweise. However, it is correct that the price the user pays is that a consistent problem will only be detected after a longer delay.

yanghua commented:
hi @tillrohrmann I have refactored this PR and counted the failure number in the `CheckpointCoordinator`. I think I should push the implementation to let you estimate. I have tested the counter's run path, but I don't know if it is the [right way](https://github.com/apache/flink/pull/6567/files#diff-a38ea0fa799bdaa0b354d80cd8368c60R1010) of failing the `ExecutionGraph` . And maybe the test case I added have too much assert, I will reduce it.

yanghua commented:
hi @tillrohrmann what do you think about the latest implementation?

azagrebin commented:
Thanks for the update, @yanghua. Looking at the checkpoint coordinator more deeply, I think we firstly have to work a bit more on design for this kind of change. We have to take into account at least the following points (roughly):

- Introduce the separate component/class responsible for failure management, counting e.g. `CheckpointFailureManager` or something

- Job manager should 
    - construct `CheckpointFailureManager`
    - configure with the max failure count and a proper action how to fail.
    - pass it to `CheckpointCoordinator`

- `CheckpointCoordinator`
    - should give callbacks to `CheckpointFailureManager` about failures and successes of checkpoints
    - needs some refactoring to distinguish better failures of `PendingCheckpoint`. Now it is `abortXXX()` methods which do not provide enough information to count it as a failure for `CheckpointFailureManager` or not (like we have for `triggerSavepoint` in `CheckpointDeclineReason`). At the end there should be clear place in the `CheckpointCoordinator` where to give callbacks to `CheckpointFailureManager`, e.g.:
        - `CheckpointDeclineReason.EXCEPTION` result of `triggerSavepoint`
        - some cases of `PendingCheckpoint.abortDeclined()`, `abortError()`, maybe `abortExpired()`

- Consider having only `DecliningCheckpointExceptionHandler` on `TaskExecutor` side and letting now to handle all failure cases only in `CheckpointCoordinator`

There might be more points. I suggest we step back and continue discussion in the jira issue. Once we have clear design, a PR can be opened again.

yanghua commented:
@azagrebin thanks for your suggestion, I agree and reconsider more details.

Considering that @tillrohrmann  has said that the current checkpoint exception handler should also be implemented in `CheckpointCoordinator`. 

> I understand why you implemented it the way you did. I think the `setFailOnCheckpointingErrors` should actually also be handled by the `CheckpointCoordinator`/`JM` and not on the Task level (but this is a different story). This looks a little bit like a shortcut we made back in the days.

Do we need to take this refactoring into account? Because this PR is actually a supplement to the checkpoint exception handler.

azagrebin commented:
> Do we need to take this refactoring into account? Because this PR is actually a supplement to the checkpoint exception handler.

Yes, I think we can have the same semantics as in the previous only-TM implementation (`setFailOnCheckpointingErrors` means `tolerableFailures == 0`) but now everything should happen in `CheckpointCoordinator`. That is why I also mentioned this point:

> Consider having only DecliningCheckpointExceptionHandler on TaskExecutor side and letting now to handle all failure cases only in CheckpointCoordinator

yanghua commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointFailureManager.java:
```diff
@@ -0,0 +1,85 @@
+		synchronized (lock) {
+			if (failOnCheckpointingErrors ||
+				continuousFailureCounter.incrementAndGet() > tolerableCpFailureNumber) {
+				executionGraph.failGlobal(new Throwable(reason));
```
hi @azagrebin and @tillrohrmann , Can you take a general look at my PR implementation and comment on how to trigger a job failure when the failure condition is met? My current implementation seems to be a bit problematic (cyclic dependency), which caused a lot of test failures.

azagrebin commented:
Hi @yanghua,

In general, `executionGraph.failGlobal` looks good to me to fail, but I think the `CheckpointFailureManager` should be constructed with a callback which will fail job. It means that `JobMaster` or `ExecutionGraphBuilder` should decide how to fail. We also have to make that checkpointing in Task Executor does not fail the job anymore, only the `CheckpointFailureManager` as a central point of failure.

Another point is that inserting failure callback into different places in `CheckpointCoordinator` close to cause looks rather invasive. Ideally, there should be 2 places where `CheckpointFailureManager` should get callback:
- in `triggerCheckpoint(long timestamp, boolean isPeriodic)` checking `CheckpointTriggerResult` (sync part)
- similar to `PendingCheckpointStats` getting callback in `PendingCheckpoint.reportFailedCheckpoint` (async part) but with clear defined cause like `CheckpointDeclineReason` for sync part.

`PendingCheckpoint` does not have such clearly defined failure handling as `triggerCheckpoint` and might need some refactoring to distinguish failure cases. 

I suggest we firstly describe the full approach in Jira issue where we also include a list of all possible failures which `CheckpointFailureManager` needs to react upon and how we define them in code. It can be also a link to some design doc.

azagrebin commented:
Hi @yanghua,
 
can this PR be closed for now and we come back to it when we have design in the Jira issue? 

We might also want to do some refactoring of failure handling in checkpoint coordinator to prepare it for integrating the error counter from this PR.

yanghua commented:
Hi @azagrebin  I am currently busy with several other PRs and hope it will be merged into Flink 1.7.0. When those PRs are fixed, I will come back and reorganize it. I think some logic of this PR is available, but the counting logic may need to be refactored. Can we put it aside for some time?

azagrebin commented:
I created FLINK-10724 to refactor failure handling in checkpoint coordinator where I believe we should firstly prepare it for better integration with the error counter we have in this PR. Once we resolve FLINK-10724, we can reopen this PR or open another one and rebase the error counter on the result of FLINK-10724 which should be simpler then.

yanghua commented:
@azagrebin OK, agree.
